### PR TITLE
cleanup of i2c_slave driver and test implementation and documentation

### LINKED ISSUE #26805: test: drivers: i2c: i2c_slave_api:
**Describe the bug**
We introduced new board to our test bench: nucleo_f091rc 
And we get status failed on test: tests/drivers/i2c/i2c_slave_api/

Looking at original commit introducing this test:
(sha1: 94d3156540d27392236044eebcb4c3a6e6065185)

>     test: i2c: Add i2c_slave_api test
>     
>     Add a new test to make I2C transfers between an I2C controller and
>     an I2C slave driver, either by:
>     - Using a virtual I2C driver
>     - Using a physical link and 2 physical controllers
>     
>     The Physical test uses :
>     - Two EEPROM Slave drivers each connected to one I2C controller
>     - Both I2C controllers also act as master and take turns writing
>     to the opposite EEPROM Slave.
>     All of this connected physically on the board.

First issue : specifically for nucleo_f091rc and also for stm32f072b_disco, virtual I2C is disabled
`CONFIG_I2C_VIRTUAL=n`
Thus Physical EEPROM should be required, whereas no eeprom is connected to the board.

I could either:
* exclude both board from this test, but the whitelist in yaml contains only those boards !! Thus test would become useless.
  This is probably a 2nd issue, I don't understand why only those 2 boards are whitelisted. 
  Other board (not only stm32) could be tested either with physical eeprom or with virtual I2C.
Note: if external hardware is required maybe a 'harness' or 'fixture' should be added

* use virtual I2C for those boards.
Unfortunately it doesn't work.
I don't think issue is specific to STM32, but simply not tested on other boards. 
Possibly this test has not been run with virtual i2C for ages ...
@henrikbrixandersen it seem you use this test ?

**To Reproduce**
1. either remove `CONFIG_I2C_VIRTUAL=n` from file nucleo_f091rc.conf, or not
2. west build -p auto -b nucleo_f091rc tests/drivers/i2c/i2c_slave_api/
3. west flash

**Expected behavior**

**Logs and console output**
*** Booting Zephyr OS build zephyr-v2.3.0-980-gb08309e9924c  ***
```
Running test suite test_eeprom_slave
===================================================================
START - test_eeprom_slave
I: Found I2C Master device I2C_1
I: Found I2C Master device I2C_2
I: Found EEPROM device EEPROM_SLAVE_0
I: Found EEPROM device EEPROM_SLAVE_1
D: i2c: slave registered
I: EEPROM EEPROM_SLAVE_0 Attached !
D: i2c: slave registered
I: EEPROM EEPROM_SLAVE_1 Attached !
I: Start full read. Master: I2C_1, address: 0x56
D: stm32_i2c_msg_write: NACK
```

**Environment (please complete the following information):**
 - OS:  linux
 - Toolchain Zephyr SDK
 - Commit SHA: dc43c298a2

### LINKED ISSUE #27303: RFC: downgrade i2c eeprom_slave driver to test
We have a [driver accessed under `CONFIG_I2C_EEPROM_SLAVE`](https://docs.zephyrproject.org/latest/reference/peripherals/i2c_eeprom_slave.html)  that implements the I2C follower (slave) driver API.  This device isn't an actual hardware driver: it's a virtual EEPROM that stores data in an instance-specific RAM buffer, with the data exposed on an I2C bus as a I2C follower device that can be controlled by another device acting as a leader (master) on that same bus.

As such it's a reasonable example of how to write an I2C follower driver, but it's not clear that it has a real use in applications: the sole in-tree reference is in the `i2c_slave_api` test.

Can anybody make a case for having this "device" available to applications?  If not, I propose to move it into the test source directory.

### PR DESCRIPTION
**Pull request by pabigot**
Closes #27303 by moving the simulated EEPROM I2C follower device into the test that uses it.

Closes #26805 by documenting how the test is intended to be used.

See individual commit messages for more details.

This commit includes material and concepts that derive from changes originally proposed by @sjg20 in  #26003.

### CODE REVIEW

**Comment by pabigot** [comment]
DNM until 2020-08-12 to allow one week for community response regarding #27303.

File: tests/drivers/i2c/i2c_slave_api/src/eeprom_slave.c [hunk 1]
```diff
@@ -11,11 +11,11 @@
 #include <errno.h>
```
Comment (sjg20) [inline tests/drivers/i2c/i2c_slave_api/src/eeprom_slave.c:11]: I'm not very good at reviewing on github yet. But I would expect to see this file added to the CMakeLists.txt file?

**Review by sjg20 — CHANGES_REQUESTED** [review]

File: tests/drivers/i2c/i2c_slave_api/src/eeprom_slave.c (hunk 1 above)
Comment (pabigot) [reply tests/drivers/i2c/i2c_slave_api/src/eeprom_slave.c:11]: Most of the Zephyr samples and tests (including this one) have this in `CMakeLists.txt`:
```
FILE(GLOB app_sources src/*.c)
```
so there's no need to list individual sources.

Comment (sjg20) [reply tests/drivers/i2c/i2c_slave_api/src/eeprom_slave.c:11]: OK ta

**Review by sjg20 — APPROVED** [review]

File: tests/drivers/i2c/i2c_slave_api/README.txt [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+  (master)) and verify that the data supplied by the other controller
+  (as bus follower) match the expected values given the content known to
+  be present on the simulated device.
+
+Transfer of commands from one bus controller to the other is
+accomplished by hardware through having the SCL (and SDA) signals
+shorted to join the two buses.
+
+Presence of this required hardware configuration is identified by the
+`i2c_bus_short` fixture.  If the buses are not connected as required,
```
Comment (erwango) [inline tests/drivers/i2c/i2c_slave_api/README.txt:27]: > `i2c_bus_short` fixture.

Shouldn't it be filtered in test yaml file then?

Comment (pabigot) [reply tests/drivers/i2c/i2c_slave_api/README.txt:27]: Yes; apparently I didn't save those changes file so it was only present as an emacs backup.

**Review by erwango — APPROVED** [review]

**Comment by pabigot** [comment]
No objections to the API change raised in #27303 or the devel or users lists.  This is ready for merge after any additional reviews deemed necessary.

**Review by carlescufi — APPROVED** [review]

**Pull request merged**